lib/semantics/check-do.cc

```diff
@@ -647,20 +645,131 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
   doContext.Check(x);
 }
 
-// C1134
-void DoChecker::Enter(const parser::CycleStmt &) {
-  if (!context_.InsideDoConstruct()) {
-    context_.Say(
-        *context_.location(), "CYCLE must be within a DO construct"_err_en_US);
+// Return the (possibly null) name of the ConstructNode
+static NamePtr MaybeGetNodeName(const ConstructNode &construct) {
+  return std::visit(
+      [&](const auto &x) { return GetConstructName(*x); }, construct);
+}
+
+template<typename A> static parser::CharBlock GetConstructPosition(const A &a) {
+  return std::get<0>(a.t).source;
+}
+
+static parser::CharBlock GetNodePosition(const ConstructNode &construct) {
+  return std::visit(
+      [&](const auto &x) { return GetConstructPosition(*x); }, construct);
+}
+
+void DoChecker::SayBadLeave(const char *stmtChecked,
+    const char *enclosingStmtType, const ConstructNode &construct) const {
+  context_
+      .Say(*context_.location(), "%s must not leave a %s statement"_err_en_US,
+          stmtChecked, enclosingStmtType)
```

Comment:
You should be able to omit the first argument to `Say`. That's what the location defaults to.

Reply:
Thanks.  I'll change that.